mcs/class/System/Internal.Cryptography/X509Pal.cs

```diff
@@ -0,0 +1,61 @@
+//
+// Copyright (c) 2018 Microsoft
+//
+// Permission is hereby granted, free of charge, to any person obtaining
+// a copy of this software and associated documentation files (the
+// "Software"), to deal in the Software without restriction, including
+// without limitation the rights to use, copy, modify, merge, publish,
+// distribute, sublicense, and/or sell copies of the Software, and to
+// permit persons to whom the Software is furnished to do so, subject to
+// the following conditions:
+// 
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+// 
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+//
+
+using System;
+using System.Security.Cryptography;
+using System.Security.Cryptography.X509Certificates;
+
+namespace Internal.Cryptography.Pal
+{
+	// Temporary until real X509Pal from CoreFX could be used.
```

Comment:
We should separate the `X500DistinguishedName` from the general `X509Certificate` code.  A general `X509Pal` class might not be possible because the code needs to be integrated with our TLS Provider framework.

As a general rule, we should ask ourselves whether or not a particular class needs to hold any native resources such as pointers to some native object.

Reply:
This basically boils down to whether we want to support only one implementation per platform or not. CoreFX strictly uses AppleTLS on macOS, OpenSSL on Linux and native implementation on Windows. Mono currently builds both BoringSSL and AppleTLS code on macOS and switches them at runtime. I have an open CoreFX issue on making the code a little bit more flexible so both options are possible (https://github.com/dotnet/corefx/issues/30606).

This was just a placeholder class to supplement whatever platform-specific implementation CoreFX uses now and default to the fully managed one. In this dependant PR - https://github.com/dotnet/corefx/pull/30572 - we agreed to drop the OpenSSL-specific code for X500 name decoding, so that part of PAL interface may soon be gone too.

Reply:
The primary reason why we need both implementations at the same time is that it would otherwise become nearly impossible to test certain things.  For instance, triggering a TLS Renegotiation is not supported in boringssl, so the only way of properly testing renegotiation in the BTLS client-side code is by using AppleTLS on the server-side.  In addition to that, I also don't want the BTLS code to bit-rot and the best way of preventing it is by having it in active use.  Most of us, including myself, are working on Macs most of the time and not on Unix/Linux - so the best way of keeping it in active use is by having it built on Mac with the option to quickly and easily switch.

Reply:
Definitely makes sense to me. I also work mostly on Mac these days and I test with both the Apple and BTLS code bases. For me a runtime switch is always a nice option, but it requires CoreFX changes. I am slowly going through process of upstreaming my local CoreFX changes. I explored this one too, but ultimately I finished it only for the X509* classes and the macOS backend and it now sits on a stale branch. I will wait for feedback from CoreFX people and revisit it later.

Reply:
Could you possibly take this out of this PR?  I have a better idea for this in my current working copy and splitting it up into pieces will make it easier.